### PR TITLE
bug/CASMTRIATGE-3310 - add cray-dns-unbound step to prereqs for upgrade to csm-1.2

### PR DESCRIPTION
**Pull request by dle-hpe**
## Summary and Scope

_Summarize what has changed. Explain why this PR is necessary. What is impacted? Is this a new feature, critical bug fix, etc?_
- add cray-dns-unbound to prerequiste.sh in the upgrade to csm-1.2
_Is this change backwards incompatible, backwards compatible, or a backwards compatible bugfix?_
- yes
## Issues and Related PRs

_List and characterize relationship to Jira/Github issues and other pull requests. Be sure to list dependencies._

CASMTRIAGE-3310

## Testing

_List the environments in which these changes were tested._

### Tested on:

- hela
  - tested with single dns forwarder
  - tested with multiple dns forwarders
### Test description:

_How were the changes tested and success verified? If schema changes were part of this change, how were those handled in your upgrade/downgrade testing?_

- Were the install/upgrade-based validation checks/tests run (goss tests/install-validation doc)?n/a
- Were continuous integration tests run? If not, why?n/a
- Was upgrade tested? If not, why?yes
- Was downgrade tested? If not, why?yes
- Were new tests (or test issues/Jiras) created for this change?no

## Risks and Mitigations

_Are there known issues with these changes? Any other special considerations?_
- none i can think of but customers that didnt properly maintain customizations.yaml could be an issue

## Pull Request Checklist

- [x] Version number(s) incremented, if applicable
- [ ] Copyrights updated
- [x] License file intact
- [ ] Target branch correct
- [ ] CHANGELOG.md updated
- [ ] Testing is appropriate and complete, if applicable
- [ ] [HPC Product Announcement](https://cray.slack.com/archives/C026TVCSXLH) prepared, if applicable

### CODE REVIEW

**Review by johren-hpe — CHANGES_REQUESTED** [review]

File: upgrade/1.2/scripts/upgrade/prerequisites.sh [hunk 1]
```diff
@@ -298,6 +298,46 @@ else
...
 fi
 
+state_name="UPGRADE_UNBOUND"
+state_recorded=$(is_state_recorded "${state_name}" $(hostname))
+if [[ $state_recorded == "0" && $(hostname) == "ncn-m001" ]]; then
+    echo "====> ${state_name} ..."
+    {
+    dns_forwarder=$(kubectl -n loftsman get secret site-init -o jsonpath='{.data.customizations\.yaml}' | base64 -d|yq r - spec.network.netstaticips.system_to_site_lookups)
+    system_name=$(kubectl -n loftsman get secret site-init -o jsonpath='{.data.customizations\.yaml}' | base64 -d|yq r - spec.network.dns.external)
+    unbound_version=$(ls ${CSM_ARTI_DIR}/helm |grep cray-dns-unbound|sed -e 's/\.[^./]*$//'|cut -d '-' -f4)
```
[the remark this answers is not in the document]
Comment (dle-hpe) [reply upgrade/1.2/scripts/upgrade/prerequisites.sh:309]: i added checks for nulls but did not add any checks for variable format

File: upgrade/1.2/scripts/upgrade/prerequisites.sh [hunk 2]
```diff
@@ -298,6 +298,46 @@ else
...
+
+
+    cat > unbound.yaml <<EOF
```
[the remark this answers is not in the document]
Comment (dle-hpe) [reply upgrade/1.2/scripts/upgrade/prerequisites.sh]: i added a variable to save the manifest folder and currently set to  /tmp

File: upgrade/1.2/scripts/upgrade/prerequisites.sh [hunk 3]
```diff
@@ -303,12 +303,21 @@ state_recorded=$(is_state_recorded "${state_name}" $(hostname))
...
     system_name=$(kubectl -n loftsman get secret site-init -o jsonpath='{.data.customizations\.yaml}' | base64 -d|yq r - spec.network.dns.external)
     unbound_version=$(ls ${CSM_ARTI_DIR}/helm |grep cray-dns-unbound|sed -e 's/\.[^./]*$//'|cut -d '-' -f4)
 
+    echo "ERROR: null value found.  See list of variables"
+    echo "dns_forwarder is $dns_forwarder."
+    echo "system_name is $system_name."
+    echo "unbound_version is $unbound_version."
 
-    cat > unbound.yaml <<EOF
+    if [ -z "$dns_forwarder" ] || [ -z "$system_name" ] || [ -z "$unbound_version" ]; then
```
Comment (johren-hpe) [inline upgrade/1.2/scripts/upgrade/prerequisites.sh:316]: This line (316) should be above line 311, correct?

Comment (johren-hpe) [reply upgrade/1.2/scripts/upgrade/prerequisites.sh:316]: Or at least 311 should be below 316.

File: upgrade/1.2/scripts/upgrade/prerequisites.sh [hunk 4]
```diff
@@ -298,6 +298,58 @@ else
...
+      global:
+        appVersion: $unbound_version
+      localZones:
+      - localType: static
+        name: local
+    version: $unbound_version
+EOF
+
+echo "$manifest_folder/unbound.yaml"
+cat /tmp/unbound.yaml
```
Comment (johren-hpe) [inline upgrade/1.2/scripts/upgrade/prerequisites.sh:343]: Minor, but if we're using $manifest_folder, then we should use it on line 343, too.

File: upgrade/1.2/scripts/upgrade/prerequisites.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (johren-hpe) [reply upgrade/1.2/scripts/upgrade/prerequisites.sh:309]: I think that's fine for now.   I'm most concerned about the case that these values are not found.

File: upgrade/1.2/scripts/upgrade/prerequisites.sh [hunk 5]
```diff
@@ -298,6 +298,58 @@ else
...
+    echo "====> ${state_name} ..."
+    {
+    manifest_folder='/tmp'
+    dns_forwarder=$(kubectl -n loftsman get secret site-init -o jsonpath='{.data.customizations\.yaml}' | base64 -d|yq r - spec.network.netstaticips.system_to_site_lookups)
+    system_name=$(kubectl -n loftsman get secret site-init -o jsonpath='{.data.customizations\.yaml}' | base64 -d|yq r - spec.network.dns.external)
+    unbound_version=$(ls ${CSM_ARTI_DIR}/helm |grep cray-dns-unbound|sed -e 's/\.[^./]*$//'|cut -d '-' -f4)
+
+
+    if [ -z "$dns_forwarder" ] || [ -z "$system_name" ] || [ -z "$unbound_version" ]; then
+    echo "ERROR: null value found.  See list of variables"
```
Comment (johren-hpe) [inline upgrade/1.2/scripts/upgrade/prerequisites.sh:313]: indentation

File: upgrade/1.2/scripts/upgrade/prerequisites.sh [hunk 6]
```diff
@@ -298,6 +298,58 @@ else
...
+      localZones:
+      - localType: static
+        name: local
+    version: $unbound_version
+EOF
+
+echo "$manifest_folder/unbound.yaml"
+cat /tmp/unbound.yaml
+
+loftsman ship --charts-path ${CSM_ARTI_DIR}/helm/ --manifest-path manifest_folder/unbound.yaml
```
Comment (johren-hpe) [inline upgrade/1.2/scripts/upgrade/prerequisites.sh:345]: manifest_folder is missing $

**Review by johren-hpe — APPROVED** [review]

**Pull request merged**